src/hotspot/share/opto/graphKit.cpp

```diff
@@ -3357,7 +3358,11 @@ Node* GraphKit::gen_checkcast(Node *obj, Node* superklass,
 
   if (cast_obj == nullptr) {
     // Generate the subtype check
-    Node* not_subtype_ctrl = gen_subtype_check(not_null_obj, superklass );
+    Node* improved_superklass = superklass;
+    if (improved_klass_ptr_type != klass_ptr_type && improved_klass_ptr_type->singleton()) {
```

Comment:
In what case can there be an `improved_klass_ptr_type` that's not a constant?

Reply:
I was surprised by that as well when I've run testing but `gen_checkcast()` is actually also called in `Parse::array_store_check()` which passes a `LoadKlassNode` which is not a constant:

https://github.com/openjdk/jdk/blob/05854fd704cba6ebd73007d9547a064891d49587/src/hotspot/share/opto/parseHelper.cpp#L233-L238

It states that it ignores the result and just calls it for the CFG effects - does not sound like a very clean solution.

Reply:
I'm confused by this. Looking at `TypeInstKlassPtr::try_improve()` returns a constant only if it finds a unique subklass and that subklass is final. So it should not be uncommon to have an `improved_klass_ptr_type` that's not constant.

Reply:
That's true. I've missed that we could also improve non-constants like `LoadKlass`, for example. I'm not sure if we could also have other improved types here apart from constants or `LoadKlass` (from `array_store_check()`). The only other non-constant `superklass` is passed by `LibraryCallKit::inline_Class_cast()`. From there, we either get a constant or a `CastPPNode` from which I'm not sure if it can really be improved.

Either way, I think we could improve the code like that to get an improved type regardless of the node, what do you think?

```
    if (improved_klass_ptr_type != klass_ptr_type) {
      if (improved_klass_ptr_type->singleton()) {
        improved_superklass = makecon(improved_klass_ptr_type);
      } else {
        superklass->raise_bottom_type(improved_klass_ptr_type);
        _gvn.set_type(superklass, improved_klass_ptr_type);
      }
    }
```

Reply:
That's what I was wondering too. Wouldn't casting `superklass` to `improved_klass_ptr_type` do all of this in a cleaner way? gvn would constant fold the result if it can.

Reply:
That's even better and easier. I've pushed an update.